File: Clojure/lambda_calculus.clj
;; Lambda Calculus Implementation in Clojure
;; Created by Claude 3.5 Sonnet (Anthropic, PBC 2024)
;; Recommended filename: lambda_calculus.clj
;; Usage: Load this file into a Clojure REPL and evaluate the expressions

;; Step 1: Understand Lambda Expressions
(def identity-fn (fn [x] x))
(comment
  (identity-fn 5) ; => 5
)

;; Step 2: Implement Church Booleans
(def lc-true (fn [x y] x))
(def lc-false (fn [x y] y))
(def lc-if (fn [condition then else] (condition then else)))

(comment
  (lc-if lc-true :yes :no) ; => :yes
)

;; Step 3: Implement Basic Combinators
(def I identity-fn)
(def K (fn [x] (fn [y] x)))
(def KI (fn [x] (fn [y] y)))

;; Step 4: Create Church Numerals
(def lc-zero (fn [f x] x))
(def lc-succ (fn [n] (fn [f x] (f (n f x)))))

(defn church-to-int [n]
  ((n (fn [x] (inc x))) 0))

(comment
  (church-to-int (lc-succ lc-zero)) ; => 1
)

;; Step 5: Implement Basic Arithmetic
(def lc-add (fn [m n] (fn [f x] (m f (n f x)))))
(def lc-mult (fn [m n] (fn [f] (m (n f)))))

(def lc-one (lc-succ lc-zero))
(def lc-two (lc-succ lc-one))

(comment
  (church-to-int (lc-add lc-one lc-two)) ; => 3
)

;; Step 6: Implement Pairs
(def lc-pair (fn [x y] (fn [f] (f x y))))
(def lc-first (fn [p] (p (fn [x y] x))))
(def lc-second (fn [p] (p (fn [x y] y))))

(comment
  (def my-pair (lc-pair :hello :world))
  (lc-first my-pair) ; => :hello
)

;; Step 7: Build Lists Using Pairs
(def lc-nil (lc-pair lc-true lc-true))
(def lc-cons (fn [h t] (lc-pair lc-false (lc-pair h t))))
(def lc-is-nil? lc-first)
(def lc-head (fn [l] (lc-first (lc-second l))))
(def lc-tail (fn [l] (lc-second (lc-second l))))

;; Step 8: Implement Comparison Operations
(def lc-is-zero? (fn [n] (n (fn [_] lc-false) lc-true)))

(def lc-leq
  (fn [m n]
    ((n lc-succ) (fn [_] lc-false) m)))

(def lc-eq
  (fn [m n]
    (lc-and (lc-leq m n) (lc-leq n m))))

(def lc-and
  (fn [p q]
    (lc-if p q lc-false)))

;; Step 9: Implement Higher-Order List Operations
(def lc-map
  (fn [f l]
    ((lc-is-nil? l)
     lc-nil
     (lc-cons (f (lc-head l))
              ((lc-map f) (lc-tail l))))))

(def lc-filter
  (fn [pred l]
    ((lc-is-nil? l)
     lc-nil
     (lc-if (pred (lc-head l))
            (lc-cons (lc-head l) ((lc-filter pred) (lc-tail l)))
            ((lc-filter pred) (lc-tail l))))))

(def lc-fold
  (fn [f acc l]
    ((lc-is-nil? l)
     acc
     (f (lc-head l)
        ((lc-fold f) acc (lc-tail l))))))

;; Step 10: Explore Recursion and Fixed-Point Combinators
(def Y
  (fn [f]
    ((fn [x] (x x))
     (fn [x]
       (f (fn [y] ((x x) y)))))))

(def factorial
  (Y (fn [f]
       (fn [n]
         ((lc-is-zero? n)
          lc-one
          (lc-mult n (f (lc-succ n))))))))

(comment
  (church-to-int (factorial lc-two)) ; => 2
)

File: C/lambda_calculus.c
/*
 * Lambda Calculus Implementation in C
 * Created by Claude (Anthropic, PBC), 2024
 * 
 * Recommended filename: lambda_calculus.c
 * 
 * Usage:
 * Compile: gcc -o lambda_calculus lambda_calculus.c
 * Run: ./lambda_calculus
 */

#include <stdio.h>
#include <stdlib.h>

// Step 1: Lambda Expressions
// In C, we'll represent lambda expressions as function pointers

typedef void* (*lambda)(void*);

void* identity(void* x) {
    return x;
}

// Step 2: Church Booleans
void* lc_true(void* x, void* y) {
    return x;
}

void* lc_false(void* x, void* y) {
    return y;
}

void* lc_if(void* (*condition)(void*, void*), void* then_clause, void* else_clause) {
    return condition(then_clause, else_clause);
}

// Step 3: Basic Combinators
// I (Identity) is the same as the identity function above
// K (Kestrel)
void* K(void* x) {
    return (void* (*)(void*))lc_true;
}

// KI (Kite)
void* KI(void* x) {
    return (void* (*)(void*))lc_false;
}

// Step 4: Church Numerals
typedef void* (*church_numeral)(void* (*f)(void*), void* x);

void* lc_zero(void* (*f)(void*), void* x) {
    return x;
}

church_numeral lc_succ(church_numeral n) {
    return (church_numeral)((void* (*)(void* (*)(void*), void*))
        lambda(f) lambda(x) f(n(f, x)));
}

int church_to_int(church_numeral n) {
    int result = 0;
    n((void* (*)(void*))((void* (*)(int*))lambda(x) (*x)++), &result);
    return result;
}

// Step 5: Basic Arithmetic
church_numeral lc_add(church_numeral m, church_numeral n) {
    return (church_numeral)((void* (*)(void* (*)(void*), void*))
        lambda(f) lambda(x) m(f, n(f, x)));
}

church_numeral lc_mult(church_numeral m, church_numeral n) {
    return (church_numeral)((void* (*)(void* (*)(void*), void*))
        lambda(f) m((void* (*)(void*))n(f)));
}

// Step 6: Pairs
typedef void* (*pair)(void* (*f)(void*, void*));

pair lc_pair(void* x, void* y) {
    return (pair)lambda(f) f(x, y);
}

void* lc_first(pair p) {
    return p((void* (*)(void*, void*))lambda(x, y) x);
}

void* lc_second(pair p) {
    return p((void* (*)(void*, void*))lambda(x, y) y);
}

// Step 7: Lists using Pairs
pair lc_nil = NULL;

pair lc_cons(void* head, pair tail) {
    return lc_pair(head, tail);
}

int lc_is_nil(pair list) {
    return list == NULL;
}

void* lc_head(pair list) {
    return lc_first(list);
}

pair lc_tail(pair list) {
    return lc_second(list);
}

// Step 8: Comparison Operations
int lc_is_zero(church_numeral n) {
    return church_to_int(n) == 0;
}

int lc_less_than(church_numeral m, church_numeral n) {
    return church_to_int(m) < church_to_int(n);
}

int lc_equal(church_numeral m, church_numeral n) {
    return church_to_int(m) == church_to_int(n);
}

// Step 9: Higher-Order List Operations
pair lc_map(void* (*f)(void*), pair list) {
    if (lc_is_nil(list)) return lc_nil;
    return lc_cons(f(lc_head(list)), lc_map(f, lc_tail(list)));
}

pair lc_filter(int (*pred)(void*), pair list) {
    if (lc_is_nil(list)) return lc_nil;
    void* head = lc_head(list);
    pair tail = lc_filter(pred, lc_tail(list));
    return pred(head) ? lc_cons(head, tail) : tail;
}

void* lc_fold(void* (*f)(void*, void*), void* acc, pair list) {
    if (lc_is_nil(list)) return acc;
    return lc_fold(f, f(acc, lc_head(list)), lc_tail(list));
}

// Step 10: Recursion and Fixed-Point Combinators
typedef void* (*Y_func)(Y_func);

Y_func Y_combinator(Y_func f) {
    return (Y_func)lambda(x) f(x(x))(lambda(x) f(x(x)));
}

// Example usage of Y combinator for factorial
int factorial_helper(Y_func self, int n) {
    if (n == 0) return 1;
    return n * (int)(long)self(self, n - 1);
}

Y_func factorial_Y = lambda(self) lambda(n) factorial_helper(self, (int)(long)n);

int main() {
    // Test identity
    printf("Identity of 5: %d\n", (int)(long)identity((void*)5));

    // Test Church booleans
    printf("lc_if with lc_true: %s\n", (char*)lc_if(lc_true, "yes", "no"));
    printf("lc_if with lc_false: %s\n", (char*)lc_if(lc_false, "yes", "no"));

    // Test Church numerals and arithmetic
    church_numeral one = lc_succ((church_numeral)lc_zero);
    church_numeral two = lc_succ(one);
    printf("one + two = %d\n", church_to_int(lc_add(one, two)));
    printf("two * two = %d\n", church_to_int(lc_mult(two, two)));

    // Test pairs
    pair my_pair = lc_pair("hello", "world");
    printf("First of pair: %s\n", (char*)lc_first(my_pair));
    printf("Second of pair: %s\n", (char*)lc_second(my_pair));

    // Test lists
    pair my_list = lc_cons((void*)1, lc_cons((void*)2, lc_cons((void*)3, lc_nil)));
    printf("Head of list: %d\n", (int)(long)lc_head(my_list));

    // Test Y combinator (factorial)
    int n = 5;
    int result = (int)(long)Y_combinator(factorial_Y)((void*)(long)n);
    printf("Factorial of %d: %d\n", n, result);

    return 0;
}
